### ISSUE TITLE
[Feature Request] Reduce `beartype.claw` warning verbosity by truncating long exception messages

### ISSUE THREAD

**Issue by tolomea** [beartype/beartype#258]
I'm trying out beartype claw on a large project with poor typing that has never had a static checker run on it. So there's a lot of just garbage in the typing.

Currently I have a lot of test fails, but running even one test produces thousands of lines of output most of it noise. For example this:

```
..\..\..\.venv\eligible310\lib\site-packages\beartype\_decor\decorcore.py:420
  C:\Users\Gordon\.venv\eligible310\lib\site-packages\beartype\_decor\decorcore.py:420: BeartypeClawDecorWarning: Function retain.other.other.soft_close() in file "C:\Users\Gordon\code\eligible\eligible\retain\other\other.py" line 330:
  function retain.other.other.soft_close() parameter "event_type" PEP 563-postponed type hint 'Event.PORTALLED | Event.BOOKED_APPOINTMENT' syntactically invalid (i.e., "unsupported operand type(s) for |: 'str' and 'str'") under:
```

is followed by 500 lines of detail.

Is there an easy way to suppress the detail so I can see the actual complaints easier?

**Comment by leycec** [comment]
> I'm trying out beartype claw on a large project...

**Wow!** That's so exciting. I can't wait to see what happens next despite being pretty sure that what happens next is `beartype.claw` causes a trainwreck.

> ...with poor typing that has never had a static checker run on it.

*...heh.* The worst-case scenario, huh? Allow me to briefly retrieve my jalapeño-coated popcorn.

![popcorn blob](https://media.tenor.com/-O21Wp3AyuwAAAAM/joker-eating-popcorn.gif)

> ...but running even one test produces thousands of lines of output most of it noise.

**So.** It's even worse than I thought, huh? Would you mind posting a bit of that warning chuff? Are we talking insanely verbose tracebacks here *or* something even more insidious? If the former, `beartype.claw` can (and really should) avoid emitting tracebacks with its warnings by default; if the latter, my name is Clueless Joe and I shrug non-committally.

Yeah... it's the former, isn't it? I sigh. :face_exhaling: 

## Here's the Plan – Bear with Me

**Okay.** I am now thinking of a new `BeartypeConf(on_decorator_exception_is_warning_traceback: bool = False)` configuration option **defaulting to `False`.** If `True`, then `beartype.claw` will append exception tracebacks to warnings like it currently does, which is making you loathe `beartype.claw` more and more with each passing `git` commit; if `False`, then `beartype.claw` will *not* do that, which will briefly subdue your loathing before the next dumpster fire. Would that resolve the calamitous trash-storm that `beartype.claw` is currently hurling at your team?

You're absolutely right, of course; if the signal-to-noise ratio is indeed as bad as you suggest,<sup>**Spoiler:** *it's even worse.*</sup> then that is demonstrably bad. `beartype.claw` is wasting everyone's precious debugging time, which is the opposite of what `beartype.claw` should be doing.

Oh – and thanks so much for trialling @beartype's newest reckless API in your awesome codebase. We'll get this right for you and your team or my username isn't @bearbrosforevah. But tonight we play video games! Raise those controllers.

**Comment by leycec** [comment]
*...lol.* Took me forever to get the name of that option right. `decorator_exception_as_warning_traceback` it is, then! :muscle:

**Comment by tolomea** [comment]
Sorry, I should have been a bit more specific, the "detail" is mostly dumps
of the entire content of global and local scope

**Comment by tolomea** [comment]
Mostly it's going well, main issues:
1: our never before checked type hints are a hot mess
2: the above verbose warnings
3: doesn't deal with imports guarded by if TYPE_CHECKING
4: mocks were interesting, but the __class__ stunt seems to work
5: it's creating infinite recursion in celery, I haven't had time to dig into this, it might be linked to union and chord and immediate mode

**Comment by leycec** [comment]
> the "detail" is mostly dumps of the entire content of global and local scope

**Ah-ha!** Right. This is @beartype's fault for failing to truncate stupidly long exception messages in `beartype.claw` warnings. I know exactly how to resolve this for you, but am currently hip-deep in real-world tedium. I acknowledge this is awful, but... if you'd like to quietly patch this away on your end, consider manually editing the following file until I manage to excrete a well-tested stable release that does this for you:

```python
# In the private "beartype/_decor/decorcore.py" file,
# please change line 393 from this...
            str(exception)
# ...into this:
            str(exception)[:160]
```

**That's it.** In theory, that *should* truncate `beartype.claw` warnings to 160 characters at most. In practice, I've tested nothing. That will probably just blow up your entire codebase in a shower of broken tests, fuming coworkers, and anger-fuelled `git` commits.

Now that I've thought about this, though, it's clear that my prior comment was totally off. What's needed is for `beartype.claw` to just truncate overly long warnings to some reasonable length. What a relief. I was afraid I'd actually have to do some real work here. Phew! :relieved:

**Comment by leycec** [comment]
**Resolved by 91ec0eb.** Thanks so much for your unwavering patience, @tolomea! The `beartype.claw` API now intelligently truncates warnings to 1,024 characters – which still might be a bit too lenient, frankly. If you'd like to preview this awesome sauce before `beartype 0.15.1` drops, consider banging out this CLI one-liner and praying for worky:

```bash
python3 -m pip install git+https://github.com/beartype/beartype.git@91ec0eb28184337070ae91ddf7cacd3a750956bf
```

:pray: :sweat: :pray:
